openedx/core/djangoapps/content/search/api.py

```diff
@@ -981,3 +983,75 @@ def generate_user_token_for_studio_search(request):
         "index_name": STUDIO_INDEX_NAME,
         "api_key": restricted_api_key,
     }
+
+
+def force_array(extra_filter: Optional[Filter]) -> list[str]:
+    """
+    Convert a filter value into a list of strings.
+
+    Strings are wrapped in a list, lists are returned as-is (cast to `list[str]`),
+    and None results in an empty list.
+    """
+    if isinstance(extra_filter, str):
+        return [extra_filter]
+    if isinstance(extra_filter, list):
+        return cast(list[str], extra_filter)
+    return []
+
+
+def fetch_block_types(extra_filter: Optional[Filter]):
+    """
+    Fetch the block types facet distribution for the search results.
+    """
+    extra_filter_formatted = force_array(extra_filter)
+
+    client = _get_meilisearch_client()
+    index = client.get_index(STUDIO_INDEX_NAME)
+
+    response = index.search(
+        "",
+        {
+            "facets": ["block_type"],
+            "filter": extra_filter_formatted,
+            "limit": 0,
+        }
+    )
+
+    return response
+
+
+def get_all_blocks_from_context(
+    context_key: str,
+    extra_attributes_to_retrieve: Optional[list[str]],
+):
+    """
+    Gets all blocks from a context key using a meilisearch search.
+    Meilisearch works with limits of 1000 maximum; ensuring we obtain all blocks
+    requires making several queries.
+    """
+    limit = 1000
+    offset = 0
+    results = []
+
+    client = _get_meilisearch_client()
+    index = client.get_index(STUDIO_INDEX_NAME)
+
+    while True:
+        response = index.search(
+            "",
+            {
+                "filter": [f'context_key = "{context_key}"'],
+                "limit": limit,
+                "offset": offset,
+                "attributesToRetrieve": ["usage_key"] + extra_attributes_to_retrieve,
+            }
+        )
+
+        hits = response["hits"]
+        if not hits:
+            break
```

Comment:
This causes an additional request to mieillisearch each time even if the result is smaller than limit amount. We should check estimatedTotal hits with limit and offset, something like this:

```python
        hits = response["hits"]
        results.extend(hits)
        if response["estimatedTotalHits"] <= limit + offset:
            break

        offset += limit
```

Reply:
Updated: https://github.com/openedx/edx-platform/pull/37818/commits/5f9b576f5d1bb3896b98f0b75e07069ff5c587ed